src/NuGet.Core/NuGet.Commands/RestoreCommand/OriginalCaseGlobalPackageFolder.cs

```diff
@@ -0,0 +1,159 @@
+                        originalCaseContext,
+                        token);
+
+                    if (installed)
+                    {
+                        _request.Log.LogMinimal(string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.Log_ConvertedPackageToOriginalCase,
+                            identity));
+                    }
+                }
+            }
+        }
+
+        public void ConvertLockFileToOriginalCase(LockFile lockFile)
+        {
+            foreach (var library in lockFile.Libraries)
+            {
+                if (library.Type != LibraryType.Package)
+                {
+                    continue;
+                }
+
+                var path = _pathResolver.GetPackageDirectory(library.Name, library.Version);
+                library.Path = PathUtility.GetPathWithForwardSlashes(path);
+            }
+        }
+
+        public void ConvertToolRestoreResultToOriginalCase(ToolRestoreResult result)
+        {
+            // Tools have their own lock files that need to be converted.
+            ConvertLockFileToOriginalCase(result.LockFile);
+
+            // The path to the tool lock file contains the tool's ID and version, which need to be
+            // converted to original case.
+            result.LockFilePath = _toolPathResolver.GetLockFilePath(
+                result.FileTargetLibrary.Name,
+                result.FileTargetLibrary.Version,
+                result.LockFileTarget.TargetFramework);
+        }
+
+        private VersionFolderPathContext GetPathContext(PackageIdentity packageIdentity, bool lowercase)
+        {
+            return new VersionFolderPathContext(
+                packageIdentity,
+                _request.PackagesDirectory,
+                lowercase,
+                _request.Log,
+                _request.PackageSaveMode,
+                _request.XmlDocFileSaveMode);
+        }
+
+        private static PackageIdentity GetPackageIdentity(RemoteMatch remoteMatch)
+        {
+            return new PackageIdentity(
+                remoteMatch.Library.Name,
+                remoteMatch.Library.Version);
+        }
+
+        private async Task CopyToAsync(RemoteMatch remoteMatch, Stream destination, CancellationToken token)
+        {
+            var library = remoteMatch.Library;
+            
+            // Try to get the package from the local repositories first. 
+            var localPackage = NuGetv3LocalRepositoryUtility.GetPackage(
+                _localRepositories,
+                library.Name,
+                library.Version);
+
+            if (localPackage != null && File.Exists(localPackage.Package.ZipPath))
+            {
+                using (var stream = File.OpenRead(localPackage.Package.ZipPath))
+                {
+                    await stream.CopyToAsync(destination, bufferSize: 4096, cancellationToken: token);
```

Comment:
From what I've learned it won't copy async unless you specify `useAsync` in stream constructor